### DISCUSSION TITLE
[Widget]Trilium-Heatmap

### DISCUSSION THREAD

**Discussion by dvai** [TriliumNext/Trilium#4213] [Show and tell]
Display the number of note edits in Trilium, just like the GitHub contributions heatmap.
https://github.com/dvai/Trilium-Heatmap
![Trilium-Heatmap](https://github.com/zadam/trilium/assets/33273040/974bb5b7-1c30-4f22-ac32-42b2f07b1091)

**Comment by SiriusXT** [comment]
I've always wanted a widget like this

**Comment by zerebos** [comment]
This is really cool! I just have a question about [runOnServer.js](https://github.com/dvai/Trilium-Heatmap/blob/main/runOnServer.js). If I add this, should I then disable `setHistoryCount`?

**Reply by zerebos** [reply]
Also, you should add this to the [Awesome Trilium](https://github.com/Nriver/awesome-trilium) list!

**Reply by dvai** [reply]
Yes，you should disable it.

**Reply by meichthys** [reply]
FYI - I've submitted a pull request to add it to Awesome Trilium: https://github.com/Nriver/awesome-trilium/pull/31

**Comment by meichthys** [comment]
This widget is so cool! Keep up the great work! I love adding it to other notes (especially my default root note) by 'Including' it 👍
![image](https://github.com/zadam/trilium/assets/10717998/c4c77838-0b7e-4410-9ca8-ba9663a1b713)

**Reply by jwhonce** [reply]
What fun! Thank you for sharing. And thanks to @zadam for creating and sharing such a great platform.

**Comment by zadam** [comment]
Hi, that's very cool. I admit I had a very similar widget like this for years, but it's been so ugly I've never shared it. This one looks much better.

I have two suggestions for querying data from the database:

* you're taking into account only the last date of modification which ignores the changes made before that. You can use "revisions" table to get older changes.
* you can use GROUP BY to avoid querying for each date separately.

Combining those two points, you can make a single query to count everything:

```sql
SELECT COUNT(*), date FROM (
	SELECT noteId, SUBSTR(utcDateModified, 0, 11) AS date FROM notes
	UNION
	SELECT DISTINCT noteId, SUBSTR(utcDateCreated, 0, 11) AS date FROM revisions
)
GROUP BY date
```

On my pretty large database the query takes 100 milliseconds, so it's likely fast enough to not have to save/cache the results in a data note.

(the query uses schema from 0.61 and won't work on 0.60 and older)

**Reply by dvai** [reply]
Thank you for the suggestion, I will improve it.

**Reply by Ivy-End** [reply]
i follow you instruction, but got the error that 'revisions' not found. haha, i have ignored the info in your () before. I am current running it on 0.60.4. Thank you for advising, I found the query speed much more fast than before.

**Comment by Nriver** [comment]
It's on fire.

![1](https://github.com/zadam/trilium/assets/6752679/e04fa8bf-b850-45c1-9fcd-7b5f10496d0b)

**Comment by Nriver** [comment]
I like the idea of releasing a `.zip` file instead of separate files and long descriptions to configure various attributes. It saves a lot of effort compared to copying and pasting, setting attributes, and so on. It's very beginner-friendly. This could also be helpful for building a Plugin Manager in the future (if luck permits).

**Reply by zerebos** [reply]
I like it as well, and I decided if my addons start to become any more complicated I'll start releasing them as `zip` files. You're right that it could be a good basis for a plugin manager for Trilium... Maybe after the 0.61 release and I'm more familiar with the codebase I'll put together a proposal

**Comment by Ivy-End** [comment]
I have modified it with my custom stylizing, and merge 3 js files into ones. Here is what it look like:

![image](https://github.com/zadam/trilium/assets/3644544/4606582b-ef52-48d8-ba25-1b51e77ca9c2)

Here is the HTML code:

```HTML
<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="UTF-8">
        <style>
            #heatmap-tooltip {
                padding: 8px;

                font-size: 13px;
                line-height: 1.5;
                text-align: center;

                color: #FFFFFF;
                background: #24292F;

                display: none;
                position: absolute;
                border-radius: 8px;
                box-sizing: border-box;
            }
        </style>
    </head>
    <body>
        <div  style="display: flex; justify-content: center; align-items: center;">
            <svg id="trilium-heatmap"></svg>
        </div>
        <div id="heatmap-tooltip"></div>
    </body>
</html>
```
And here is my js code: **The original version clicked on each square to jump to the notes for the day, but there were times when there were no notes for the day, so that's been fixed over here.**

```JavaScript
/******************** Custom Parameters ********************/
const dataRange = [1, 32];
const colorRange = ["#9be9a8", "#216e39"];

const displayWeekTable = ['Mon', 'Wed', 'Fri'];
const displayMonthTable = [ 'Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec' ];

/******************** Drawing Parameters ********************/
const borderMargin = 16; const titleMargin = 8;
const monthBoxSize = 13; const boxSize = 13; const boxMargin = 3;
const weekBoxSize = boxSize * 3; 

const svgWidth = borderMargin * 2 + weekBoxSize + titleMargin + 52 * boxSize + 51 * boxMargin;
const svgHeight = borderMargin * 2 + monthBoxSize + titleMargin + 7 * boxSize + 6 * boxMargin;

/******************** Drawing Handles ********************/
const svg = d3.select('#trilium-heatmap').attr('width', svgWidth).attr('height', svgHeight);

const drawWeeks = svg.append('g').attr(
    'transform', `translate(${borderMargin + titleMargin}, ${borderMargin + monthBoxSize + titleMargin})`
);
const drawMonths = svg.append('g').attr(
    'transform', `translate(${borderMargin + weekBoxSize + titleMargin}, ${borderMargin})`
);
const drawBoxes = svg.append('g').attr(
    'transform', `translate(${borderMargin + weekBoxSize + titleMargin}, ${borderMargin + monthBoxSize + titleMargin})`
);

/******************** Dataset ********************/
const notesData = {};
    
const notes = await api.runOnBackend(async() => {
    const results = api.sql.getRows(`SELECT date as notesDate, COUNT(*) as notesCount FROM (
            SELECT noteId, SUBSTR(utcDateModified, 0, 11) AS date FROM notes
            UNION
            SELECT DISTINCT noteId, SUBSTR(utcDateCreated, 0, 11) AS date FROM notes
        )
        GROUP by date`);
    return results;
});

for (let i = 0; i < notes.length; i++) {
    notesData[notes[i]['notesDate']] = notes[i]['notesCount']
}

/******************** Drawing ********************/
const totalDays = new Date(new Date().getFullYear(), 1, 29).getDate() === 29 ? 366 : 365;
const colorScale = d3.scaleSequential().domain(dataRange).interpolator(d3.interpolateRgb(colorRange[0], colorRange[1]));

const startDate = new Date();
startDate.setDate(new Date().getDate() - totalDays + 1);
const startDayOfWeek = startDate.getDay();

// Draw Weeks
for (let i = 0; i < 3; i++) {
    drawWeeks.append('text').text(displayWeekTable[i])
        .attr('week', displayWeekTable[i])
        .attr('font-size', '13px')
        .attr('fill', '#1F2328')
        .attr('x', 0)
        .attr('y', (boxSize + boxMargin) * (2 * i + 1) + boxSize);
}

// Draw Boxes
let currentX = 0; let currentY = startDayOfWeek;
for (let i = totalDays - 1; i >= 0; i--) {
    let referDate = new Date();
    referDate.setDate(referDate.getDate() - i);
    drawBoxes.append('rect').attr('date', referDate.toISOString().slice(0, 10))
        .attr('width', boxSize).attr('height', boxSize)
        .attr('rx', boxMargin).attr('ry', boxMargin)
        .attr('x', (boxSize + boxMargin) * currentX).attr('y', (boxSize + boxMargin) * currentY)
        .attr('fill', () => {
                const referDateFormatted = referDate.toISOString().slice(0, 10);
                const notesCount = parseInt(notesData[referDateFormatted]);
                if (isNaN(notesCount)) {
                    return '#ebedf0';
                } else if (notesCount <= dataRange[1]) {
                    return colorScale(notesCount - 1);
                } else {
                    return colorRange[1];
                }
            })
        .on('mouseover', function(event) {
                const toolTips = document.getElementById('heatmap-tooltip');
                const heatMapDiv = document.getElementById('heatmap-tooltip').parentElement;
        
                const referDateFormatted = this.getAttribute('date');
                const notesCount = parseInt(notesData[referDateFormatted]);
        
                let bounds = heatMapDiv.getBoundingClientRect();
                
                d3.select(this).style('stroke', 'black').style('stroke-width', 0.75);
                console.log(event.clientX + " " + event.clientY);
                toolTips.innerHTML  = `${!isNaN(notesCount) ? notesCount.toString() : 'No'} notes modified in ${referDateFormatted}`;
                toolTips.style.left    = `${event.clientX - bounds.left - 96}px`;
                toolTips.style.top     = `${event.clientY - bounds.top + 12}px`;
                toolTips.style.display = 'block';
            })
        .on('mouseout', function() {
                const toolTips = document.getElementById('heatmap-tooltip');
        
                d3.select(this).style("stroke", null); 
                toolTips.style.display = 'none';
            })
        .on("click", async function() {
                const referDateFormatted = this.getAttribute('date');
                const notesCount = parseInt(notesData[referDateFormatted]);
        
                if (!isNaN(notesCount)) {
                    const noteId = await api.runOnBackend(async (referDateFormatted) => {
                        const note = await api.getDayNote(referDateFormatted);
                        return note.noteId;
                    }, [referDateFormatted]);
                    await api.waitUntilSynced();
                    api.activateNote(noteId);
                }
            });
  
    if (referDate.getDate() == 1) {
        drawMonths.append('text').text(displayMonthTable[referDate.getMonth()])
            .attr('month', displayMonthTable[referDate.getMonth()])
            .attr('font-size', '13px')
            .attr('fill', '#1F2328')
            .attr('x', (boxSize + boxMargin) * currentX)
            .attr('y', monthBoxSize);
    }
    
    currentY = (currentY + 1) % 7;
    currentX = currentX + (currentY == 0);
}

// Draw Months
for (let i = 0; i < totalDays; i++) {
    let referDate = new Date();
    referDate.setDate(referDate.getDate() - i);
    
    currentY = (currentY + 1) % 7;
    currentX = currentX + (currentY == 0);
}
```

**Comment by sottey** [comment]
I like the idea of a plugin system as well. The possible tasks that are taken for installing an extension currently are:
- Creating notes
    - Templates
    - Text Notes
    - JS Notes
    - CSS Notes
    - Etc.
- Assigning attributes to certain notes, both existing notes and imported/created notes
    - Relations (with validation that the related note exists)
    - Labels
- Error checking to roll back install (could use transaction api)
- Adding in of LaunchBar icons
- SQL direct actions (?)

Some of these arepotentially  super dangerous:
- any changes to existing data
- SQL actions
- Adding labels could be risky as it could share out unintended pages innocently or maliciously
- adding notes with #appCss could interfere with other CSS
- JS pages could potentially introduce malware/damage into the instance of Trilium

A lot of these issues could be mitigated with a bit of coding or some minor tweaks to Trilium (example: the ability to limit the tree scope of #appCss would be amazing. i.e. #appCss=2 results in the Css only being applied at it's own level and it's parent level.

I think this could be a really fun project. If anyone starts working on this I would be happy to help.

**Reply by zerebos** [reply]
If zadam doesn't get to it, I probably will work on a plugin manager of some kind at some point. I have a lot of experience making addon systems involving plugins and themes (I'm the developer of [BetterDiscord](https://github.com/BetterDiscord/BetterDiscord)). Definitely agree with the danger, some are easily mitigated, other less so.

`#appCss` could definitely scoped in that way, but it might not be necessary depending on the plugin. It might make more sense to keep it internal to the addon for things like widgets since they have specific css blocks internally.

**Reply by MorganeLeCaignec** [reply]
I've made a couple little things for trilium (hexmap and timeline widgets) and also started making .zip exports as an easier way of installing the widget, that way the whole bundle is ready as-is and can include example notes with pre-configured attributes and some documentation/explanation.

I would love a plugin manager of some kind, an easy way of bundling a plugin (with, as you say, easy ways if including styling that doesn't apply outside of the widget, etc), and ways for people to easily find and install them.

**Reply by meichthys** [reply]
@QuentinLeCaignec Until a plugin manager happens, could you add your widgets/scripts/etc to [Awesome-Trilium](https://github.com/Nriver/awesome-trilium)? I've been trying to keep that repo up to date with any snippes or trilium-based tools that could be useful.

**Reply by zerebos** [reply]
@QuentinLeCaignec as for the part about bundling and the developer experience, I've been working on some tools to help with just that, including a `@types` package for typescript use or just for autocomplete in regular JS. As well as a way to automatically update the note in Trilium with the newly built bundle.

**Reply by MorganeLeCaignec** [reply]
@meichthys I'm pleasantly surprised to learn my two widgets already are on there! I hadn't heard of this initiative, very nice. They are the trilium-hexmap and trilium-timeline widgets, linked to my Gitlab repos.

**Reply by meichthys** [reply]
Haha, cool 🎉
I reallly like the Timeline widget - i didn't realize that was yours. Kudos!

**Reply by mm21** [reply]
I'm planning to build a tool like that in [TriliumAlchemy](https://github.com/mm21/trilium-alchemy/), but working on a tool to synchronize Trilium with a local folder first. I'd like to support a plugin config file which can take a list of paths to .zip files or repo URLs which are cloned and managed by the tool. The existing functionality of the project provides a high-level interface to Trilium which makes working with attributes and child notes easy.

**Comment by spirit1431007** [comment]
very cool!
